titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/DefaultGrpcObjectsCache.java

```diff
@@ -0,0 +1,128 @@
+import com.netflix.titus.grpc.protogen.Job;
+import com.netflix.titus.grpc.protogen.Task;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class DefaultGrpcObjectsCache implements GrpcObjectsCache {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultGrpcObjectsCache.class);
+
+    protected ProtobufCache<com.netflix.titus.api.jobmanager.model.job.Job<?>, Job> jobCache;
+    protected ProtobufCache<com.netflix.titus.api.jobmanager.model.job.Task, Task> taskCache;
+
+    private final V3JobOperations jobOperations;
+    private final GrpcObjectsCacheConfiguration configuration;
+    private final LogStorageInfo<com.netflix.titus.api.jobmanager.model.job.Task> logStorageInfo;
+    private final TitusRuntime titusRuntime;
+
+    @Inject
+    public DefaultGrpcObjectsCache(V3JobOperations jobOperations,
+                                   GrpcObjectsCacheConfiguration configuration,
+                                   LogStorageInfo<com.netflix.titus.api.jobmanager.model.job.Task> logStorageInfo,
+                                   TitusRuntime titusRuntime) {
+        this.titusRuntime = titusRuntime;
+        this.configuration = configuration;
+        this.logStorageInfo = logStorageInfo;
+        this.jobOperations = jobOperations;
+    }
+
+    @Activator
+    public void activate() {
+        // Initialize cache
+        this.jobCache = new ProtobufCache<>(
+                "jobs",
+                this::toGrpcJob,
+                () -> {
+                    List<com.netflix.titus.api.jobmanager.model.job.Job> allJobs = jobOperations.getJobs();
+                    Set<String> knownJobIds = new HashSet<>();
+                    allJobs.forEach(job -> knownJobIds.add(job.getId()));
+                    return job -> !knownJobIds.contains(job.getId());
+                },
+                configuration,
+                titusRuntime
+        );
+
+        this.taskCache = new ProtobufCache<>(
+                "tasks",
+                this::toGrpcTask,
+                () -> {
+                    List<com.netflix.titus.api.jobmanager.model.job.Task> allTasks = jobOperations.getTasks();
+                    Set<String> knownTasksIds = new HashSet<>();
+                    allTasks.forEach(task -> knownTasksIds.add(task.getId()));
+                    return task -> !knownTasksIds.contains(task.getId());
+                },
+                configuration,
+                titusRuntime
+        );
+    }
+
+    @Deactivator
+    @PreDestroy
+    public void shutdown() {
+        Evaluators.acceptNotNull(jobCache, ProtobufCache::shutdown);
+        Evaluators.acceptNotNull(taskCache, ProtobufCache::shutdown);
+    }
+
+    public Job getJob(com.netflix.titus.api.jobmanager.model.job.Job<?> coreJob) {
+        if (configuration.isGrpcObjectsCacheEnabled()) {
+            return jobCache.get(coreJob.getId(), coreJob);
+        } else {
+            return toGrpcJob(coreJob);
```

Comment:
If the object cache is disabled, this method invocation would trigger serialization of the `job` object every time `getJob` is called for the `coreJob`. Should this be optimized?